### PR TITLE
Add all_reduce_coalesced functional collective

### PR DESCRIPTION
**Pull request by kumpera**
Inductor codegen is suboptimal when calling all_reduce_coalesced with input args. We need to fix inductor's calling convention for that, or something else.

Might not work if any outputs is unused.

Test code:

```python
import torch
import torch.distributed as dist
import torch.nn.functional as F
from functorch import make_fx
import os

import torch.distributed._functional_collectives as ft_c
from torch.testing._internal.common_distributed import (
    spawn_threads_and_init_comms,
)
from torch._inductor.compile_fx import compile_fx_inner

def my_fun(a, b):
    c = a * 3
    tensors = ft_c.all_reduce_coalesced([a, c, b], "sum", [0])
    return ((tensors[1] + tensors[0] + tensors[2]).sum(), )

@spawn_threads_and_init_comms(world_size=1)
def inductor_main(self):

    x = torch.arange(4).cuda() * (dist.get_rank() + 1)
    y = torch.arange(4).cuda() * (dist.get_rank() + 1)
    x = x.to(torch.float)
    y = y.to(torch.float) * 0.5
    res = make_fx(my_fun)(x, y)
    print(f"fx graph:\n{res.graph}")
    ind = compile_fx_inner(res, [x, y])
    print(f"inductor done:\n{ind}")

os.environ["PROXY_TENSOR_TRACING"] = "1"
os.environ["TORCH_COMPILE_DEBUG"] = "1"
torch._dynamo.config.output_code = True

if __name__ == "__main__":
    inductor_main(None)
```

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -14726,6 +14726,13 @@
   dispatch:
     CompositeExplicitAutograd: wait_tensor
 
+- func: all_reduce_coalesced(Tensor[] self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor[]
```
Comment (wanchaol) [inline aten/src/ATen/native/native_functions.yaml:14729]: I know that this might not be ready for review, but would like to have some suggestion about the naming. Given that we are in the aten op world, shall we make this an overload of `all_reduce` op instead of making them separate ops? i.e. I feel we can do `all_reduce.coalesced` or `all_reduce.grouped`

Comment (kumpera) [reply aten/src/ATen/native/native_functions.yaml:14729]: IDK, this will break a lot of our existing passes as they don't expect different ops with the same name.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -14720,6 +14720,13 @@
...
+  # This should be changed to distributed but it requires changes all over the place to work
+  python_module: nn
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:14725]: Since we're adding more and more of these, can we do this please?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -14720,6 +14720,13 @@
   dispatch:
     CompositeExplicitAutograd: wait_tensor
 
+- func: all_reduce_coalesced(Tensor[] self, str reduceOp, str tag, int[] ranks, int group_size) -> Tensor[]
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:14723]: Why do you make this a native function? You don't actually have any implementation for it here.
You can DEF new implementations with [torch.library](https://pytorch.org/docs/stable/library.html?highlight=torch+library#torch.library.Library). You can then do the exact same thing as you do here for it (even easily add it to your distributed namespace). And adding it to library makes it an op so it can live in the inductor graph the exact same way.

**Review by albanD — COMMENTED** [review]
I think this could be significantly simplified by making it python only!

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (kumpera) [reply aten/src/ATen/native/native_functions.yaml:14723]: IIRC, we ran into issues when defining it using `torch.library`.
I'll have to try it in a separate PR to refresh my memory.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (kumpera) [reply aten/src/ATen/native/native_functions.yaml:14725]: Ok, doing it on a separate PR to this one.

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (kumpera) [reply aten/src/ATen/native/native_functions.yaml:14723]: Another reason is that we want to move the implementation to C++ eventually as we need to support c++ backends

Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:14723]: You can still register the impl from c++ for a function that was defined in python :D

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (kumpera) [reply aten/src/ATen/native/native_functions.yaml:14725]: PR for fixing this: https://github.com/pytorch/pytorch/pull/97793

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:14723]: You mean in an export context? Why not? You can still provide a C++ implementation to your custom no python runner.

**Review by fegin — APPROVED** [review]

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (wconstab) [reply aten/src/ATen/native/native_functions.yaml:14739]: i think the reason we did this was to have a wya to get codegen for backwards from derivatives.yaml.

since that time we (gave up? postponed?) implementing backwards.  If we don't need codegen for new plan for backwards, i'm good with moving out of native and into py.

cc @kumpera

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -3984,6 +3984,123 @@ def codegen(self, wrapper):
...
+            f"{self.get_name()} = {self.inputs[0].get_name()}{self.index}"
+        )
+
+
+class ForceInPlace(ExternKernel):
+    """
+    Helper OP to encode an in/out argument that tries to make it inplace whenever possible.
+    Wrap the input of your inplace op to enable this behavior.
+
+    TODO: We should test whether wait_tensor can be a victim of reordering and lead to data races.
```
Comment (wconstab) [inline torch/_inductor/ir.py:4004]: you should at least test that in the simple cases this does not happen.  See my tests in `test_inductor_collectives.py` which use FileCheck to inspect the relative order of nodes in output codegen.

Comment (kumpera) [reply torch/_inductor/ir.py:4004]: This comment is actually stale as the final design is to emit wait_tensor calls for all of the resulting tensors

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (kumpera) [reply aten/src/ATen/native/native_functions.yaml:14739]: Alban helped me figure out how to get dispatch to work as we need from python.

I guess we down to figuring out what to do with torch::deploy before moving on with this - instead of going the opposite direction of moving the ops implementations to c++

File: torch/_inductor/ir.py [hunk 5]
```diff
@@ -3984,6 +3984,123 @@ def codegen(self, wrapper):
...
+        reduce_op: str,
+        tag: str,
+        ranks: List[int],
+        group_size: int,
+    ):
+        res = []
+
+        def wrap_input(var):
+            nonlocal res
+            op = ForceInPlace(
```
[the remark this answers is not in the document]
Comment (kumpera) [reply torch/_inductor/ir.py:4044]: I think it should work for the other collectives and allow us to remove the special case for AllReduce from scheduler.
I'll do this in a follow up PR.

File: torch/_inductor/scheduler.py [hunk 6]
```diff
@@ -213,7 +213,7 @@ def allocate(self):
                 # o what have i done.  lets make this an api
                 or (
                     isinstance(self, ExternKernelSchedulerNode)
-                    and isinstance(self.node, ir.AllReduce)
+                    and isinstance(self.node, (ir.AllReduce, ir.ForceInPlace))
```
Comment (wconstab) [inline torch/_inductor/scheduler.py:216]: hope to remove the allreduce case after unifying other collectives w/ using ForceInPlace (and renaming it) - ok in future PR
